pkg/rhcos/release/release.go

```diff
@@ -0,0 +1,157 @@
+	}
+
+	sys := &types.SystemContext{}
+	imageSource, err := reference.NewImageSource(ctx, sys)
+	if err != nil {
+		return nil, nil, err
+	}
+	defer func() {
+		if err != nil {
+			imageSource.Close()
+		}
+	}()
+
+	unparsed := image.UnparsedInstance(imageSource, nil)
+	// TODO: signature validation
+
+	img, err := image.FromUnparsedImage(ctx, sys, unparsed)
+	return imageSource, img, err
+}
+
+func imageReferences(ctx context.Context, releaseImagePullSpec string) ([]byte, error) {
+	imageSource, image, err := pull(ctx, releaseImagePullSpec)
+	if err != nil {
+		return nil, err
+	}
+	defer imageSource.Close()
+
+	layers := image.LayerInfos()
+	for i := len(layers) - 1; i >= 0; i-- {
+		layer := layers[i]
+		blob, _, err := imageSource.GetBlob(ctx, layer, blobinfocache.NoCache)
+		if err != nil {
+			return nil, err
+		}
+		defer blob.Close()
+
+		// fmt.Println(layer.MediaType) // TODO: blank from quay?  Just assume this is a gzipped tar?
+		gzipReader, err := gzip.NewReader(blob)
+		if err != nil {
+			return nil, err
+		}
+
+		tarReader := tar.NewReader(gzipReader)
+		for {
+			hdr, err := tarReader.Next()
+			if err == io.EOF {
+				break
+			}
+			if err != nil {
+				return nil, err
+			}
+
+			if hdr.Name == "release-manifests/image-references" {
+				return ioutil.ReadAll(tarReader)
+			}
+		}
+	}
+
+	return nil, errors.Errorf("no image reference found in %s", transports.ImageName(image.Reference()))
+}
+
+func osContentPullSpec(imageReferences []byte) (string, error) {
+	scheme := runtime.NewScheme()
+	imagev1.Install(scheme)
+	decoder := serializer.NewCodecFactory(scheme).UniversalDecoder(
+		imagev1.GroupVersion,
+	)
+	obj, _, err := decoder.Decode(imageReferences, nil, nil)
+	if err != nil {
+		return "", err
+	}
+
+	imageStream, ok := obj.(*imagev1.ImageStream)
+	if !ok {
+		return "", errors.Errorf("image references is a %q, not an image stream", obj.GetObjectKind())
+	}
+
+	for _, tag := range imageStream.Spec.Tags {
+		if tag.Name == "machine-os-content" {
+			if tag.From == nil {
+				return "", errors.New("machine-os-content tag has an empty 'from' property")
+			}
+
+			if tag.From.Kind != "DockerImage" {
+				return "", errors.Errorf("unrecognized machine-os-content tag from kind %q", tag.From.Kind)
+			}
+
+			return tag.From.Name, nil
+		}
+	}
+
+	return "", errors.New("no machine-os-content tag found")
+}
+
+// RHCOSBuild extracts the RHCOS build version from an OpenShift
+// release image.
+func RHCOSBuild(ctx context.Context, releaseImagePullSpec string) (string, error) {
+	imageReferencesBytes, err := imageReferences(ctx, releaseImagePullSpec)
+	if err != nil {
+		return "", err
+	}
+
+	pullSpec, err := osContentPullSpec(imageReferencesBytes)
+	if err != nil {
+		return "", err
+	}
+
+	imageSource, image, err := pull(ctx, pullSpec)
+	if err != nil {
+		return "", err
+	}
+	defer imageSource.Close()
+
+	inspect, err := image.Inspect(ctx)
+	if err != nil {
+		return "", err
+	}
+
+	version, ok := inspect.Labels["version"]
```

Comment:
Should this be "io.openshift.release.coreos-boot-image"?

Reply:
> Should this be "io.openshift.release.coreos-boot-image"?

Maybe eventually?  I think machine-os-content needs better labeling.  But at the moment:

```console
$ PULLSPEC="$(oc adm release info --image-for=machine-os-content quay.io/openshift-release-dev/ocp-release:4.0.0-0.7)"
$ oc image info "${PULLSPEC}"
Name:       quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:399582f711226ab1a0e76d8928ec55436dea9f8dc60976c10790d308b9d92181
Media Type: application/vnd.docker.distribution.manifest.v1+prettyjws
Created:    12d ago
Image Size: 0B
OS:         linux
Arch:       amd64
Entrypoint: /noentry
Labels:     com.coreos.ostree-commit=ecf8e84a7671e9954417d57fbdd62ab5c8eadebc268770b9d1e1fb5ff80f1624
            version=47.330
```

is all we have to work with.

Reply:
Remember the people maintaining the pipeline are right here and are happy to make changes or even accept patches!

What specifically do you want here?

Reply:
> What specifically do you want here?

Something like wking/coreos-assembler@738d918, but it would need something like openshift/origin#21786 to translate back from [OCI annotations][1] to the ones OpenShift uses internally.

[1]: https://github.com/opencontainers/image-spec/blob/v1.0.1/annotations.md#back-compatibility-with-label-schema